Core/include/Acts/Vertexing/ImpactPointEstimator.ipp

```diff
@@ -241,13 +241,65 @@ Acts::ImpactPointEstimator<input_track_t, propagator_t, propagator_options_t>::
   // Reference point R
   Vector3 refPoint = trkParams.referenceSurface().center(gctx);
 
-  // Perigee parameters (parameters of 2D PCA)
+  // Extract charge-related particle parameters
+  double absoluteCharge = trkParams.particleHypothesis().absoluteCharge();
+  double qOvP = trkParams.parameters()[BoundIndices::eBoundQOverP];
+
+  // Z-component of the B field at the reference position.
+  // Note that we assume a constant B field here!
+  auto fieldRes = m_cfg.bField->getField(refPoint, state.fieldCache);
+  if (!fieldRes.ok()) {
+    return fieldRes.error();
+  }
+  double bZ = (*fieldRes)[eZ];
+
+  // The particle moves on a straight trajectory if its charge is 0 or if there
+  // is no B field. In that case, the 3D PCA can be calculated analytically, see
+  // Sec 3.2 of the reference.
+  if (absoluteCharge == 0. or bZ == 0.) {
+    // Momentum direction (constant for straight tracks)
+    Vector3 momDirStraightTrack = trkParams.direction();
+
+    // Current position on the track
+    Vector3 positionOnTrack = trkParams.position(gctx);
+
+    // Distance between positionOnTrack and the 3D PCA
+    ActsScalar distanceToPca =
+        (vtxPos.template head<3>() - positionOnTrack).dot(momDirStraightTrack);
+
+    // 3D PCA
+    ActsVector<nDim> pcaStraightTrack;
+    pcaStraightTrack.template head<3>() =
+        positionOnTrack + distanceToPca * momDirStraightTrack;
+    if constexpr (nDim == 4) {
+      // Track time at positionOnTrack
+      double timeOnTrack = trkParams.parameters()[BoundIndices::eBoundTime];
+
+      ActsScalar m0 = trkParams.particleHypothesis().mass();
+      ActsScalar p = std::abs(absoluteCharge / qOvP);
```

Comment:
the particle hypothesis can extract the momentum in case of charged and uncharged particles